src/lib/sort/sort-header.ts

```diff
@@ -140,9 +247,49 @@ export class MatSortHeader implements MatSortable {
     this._rerenderSubscription.unsubscribe();
   }
 
+  /** Triggers the sort on this sort header and removes the indicator hint. */
+  _handleClick() {
+    this._sort.sort(this);
+
+    // Do not show the animation if the header was already shown in the right position.
+    if (this.viewState.toState === 'peek' || this.viewState.toState === 'active') {
+      this._disableViewStateAnimation = true;
+    }
+
+    // If the arrow is now sorted, animate the arrow into place. Otherwise, animate it away into
+    // the direction it is facing.
+    this.viewState = this._isSorted() ?
+        {fromState: this._arrowDirection, toState: 'active'} :
+        {fromState: 'active', toState: this._arrowDirection};
+
+    this._showIndicatorHint = false;
+  }
+
   /** Whether this MatSortHeader is currently sorted in either ascending or descending order. */
   _isSorted() {
     return this._sort.active == this.id &&
         (this._sort.direction === 'asc' || this._sort.direction === 'desc');
   }
+
+  /** Returns the animation state for the arrow direction (indicator and pointers). */
+  _getArrowDirectionState() {
+    return `${this._isSorted() ? 'active-' : ''}${this._arrowDirection}`;
+  }
+
+  /** Returns the arrow position state (opacity, translation). */
+  _getArrowViewState() {
+    const fromState = this.viewState.fromState;
+    return (fromState ? `${fromState}-to-` : '') + this.viewState.toState;
+  }
+
+  /**
+   * Updates the direction the arrow should be pointing. If it is not sorted, the arrow should be
+   * facing the start direction. Otherwise if it is sorted, the arrow should point in the currently
+   * active sorted direction.
+   */
+  _updateArrowDirection() {
```

Comment:
Why have an update function instead of a function like `_getArrowDirection`? The problem with the update function is that you can forget to call it, or call it when it's not necessary.

Reply:
I tried going that route but this ended up being much easier to control when the arrow changes directions.

Best example is when the sort goes from active to deactivated. By default the order for sort will be `none - asc - desc`. When the sort goes from `desc` to `none`, the arrow direction should remain the same until the animation completes (as it flies away). Only when the arrow is displayed again should it switch to showing itself as `asc` (hint is shown or activated). With just a "getter", we would probably have to rely on knowing if the arrow is animating away and have some additional logic about making it keep the "old direction" in the meantime.

Reply:
Expand comment to explain this, then?